File: app/services/venue_service.py
from typing import List, Optional, Tuple

from sqlalchemy import func
from sqlalchemy.orm import Session, joinedload

from app.models import User, Venue, VenueStaff
from app.models.venue_staff import VenueRole
from app.schemas.venue import VenueCreate, VenueStaffCreate, VenueStaffUpdate, VenueUpdate


class VenueService:
    """
    Service for managing venues and venue staff.
    """

    @staticmethod
    def create_venue(db: Session, venue_data: VenueCreate, owner: User) -> Venue:
        """
        Create a new venue and assign the creating user as owner.
        """
        venue = Venue(**venue_data.model_dump())
        db.add(venue)
        db.flush()

        venue_staff = VenueStaff(venue_id=venue.id, user_id=owner.id, role=VenueRole.OWNER)
        db.add(venue_staff)

        db.commit()
        db.refresh(venue)
        return venue

    @staticmethod
    def update_venue(db: Session, venue: Venue, venue_data: VenueUpdate) -> Venue:
        """
        Update venue details.
        """
        update_data = venue_data.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(venue, field, value)

        db.commit()
        db.refresh(venue)
        return venue

    @staticmethod
    def delete_venue(db: Session, venue: Venue) -> None:
        """
        Delete a venue and all associated data.
        """
        db.delete(venue)
        db.commit()

    @staticmethod
    def add_venue_staff(db: Session, venue: Venue, user: User, staff_data: VenueStaffCreate) -> VenueStaff:
        """
        Add a staff member to a venue.
        """
        venue_staff = VenueStaff(venue_id=venue.id, user_id=user.id, role=staff_data.role)
        db.add(venue_staff)
        db.commit()
        db.refresh(venue_staff)
        return venue_staff

    @staticmethod
    def update_venue_staff(db: Session, venue_staff: VenueStaff, staff_data: VenueStaffUpdate) -> VenueStaff:
        """
        Update staff member details.
        """
        update_data = staff_data.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(venue_staff, field, value)

        db.commit()
        db.refresh(venue_staff)
        return venue_staff

    @staticmethod
    def remove_venue_staff(db: Session, venue_staff: VenueStaff) -> None:
        """
        Remove a staff member from a venue.
        """
        db.delete(venue_staff)
        db.commit()

    @staticmethod
    def get_venue_staff(db: Session, venue: Venue) -> List[VenueStaff]:
        """
        Get all staff members for a venue.
        """
        return (
            db.query(VenueStaff)
            .options(joinedload(VenueStaff.user))
            .filter(VenueStaff.venue_id == venue.id)
            .order_by(VenueStaff.role, VenueStaff.joined_at)
            .all()
        )

    @staticmethod
    def get_user_venues(db: Session, user: User) -> List[Venue]:
        """
        Get all venues where a user is a staff member.
        """
        return (
            db.query(Venue)
            .join(VenueStaff)
            .filter(VenueStaff.user_id == user.id)
            .order_by(VenueStaff.role, Venue.name)
            .all()
        )

    @staticmethod
    def user_can_manage_venue(db: Session, user: User, venue: Venue) -> bool:
        """
        Check if a user has management permissions for a venue.
        """
        venue_staff = (
            db.query(VenueStaff)
            .filter(
                VenueStaff.venue_id == venue.id,
                VenueStaff.user_id == user.id,
                VenueStaff.role.in_([VenueRole.OWNER, VenueRole.MANAGER]),
            )
            .first()
        )
        return venue_staff is not None

    @staticmethod
    def user_is_venue_owner(db: Session, user: User, venue: Venue) -> bool:
        """
        Check if a user is an owner of a venue.
        """
        venue_staff = (
            db.query(VenueStaff)
            .filter(
                VenueStaff.venue_id == venue.id,
                VenueStaff.user_id == user.id,
                VenueStaff.role == VenueRole.OWNER,
            )
            .first()
        )
        return venue_staff is not None

    @staticmethod
    def list_venues(
        db: Session,
        city: Optional[str] = None,
        state: Optional[str] = None,
        has_sound_provided: Optional[bool] = None,
        has_parking: Optional[bool] = None,
        min_capacity: Optional[int] = None,
        max_capacity: Optional[int] = None,
        skip: int = 0,
        limit: int = 100,
    ) -> Tuple[List[Venue], int]:
        """
        List venues with optional filters.
        """
        query = db.query(Venue)

        if city is not None:
            query = query.filter(func.lower(Venue.city) == func.lower(city))

        if state is not None:
            query = query.filter(func.upper(Venue.state) == func.upper(state))

        if has_sound_provided is not None:
            query = query.filter(Venue.has_sound_provided == has_sound_provided)

        if has_parking is not None:
            query = query.filter(Venue.has_parking == has_parking)

        if min_capacity is not None:
            query = query.filter(Venue.capacity >= min_capacity)

        if max_capacity is not None:
            query = query.filter(Venue.capacity <= max_capacity)

        total = query.count()
        venues = query.order_by(Venue.name).offset(skip).limit(limit).all()

        return venues, total

File: src/services/venueService.js
import { apiClient } from '../utils/apiClient';

const API_BASE_URL = process.env.REACT_APP_API_URL || "http://127.0.0.1:8000/api/v1";

export const venueService = {
  async createVenue(venueData, imageFile = null) {
    const formData = new FormData();
    
    // Add all venue data fields to FormData
    formData.append("name", venueData.name);
    if (venueData.description) {
      formData.append("description", venueData.description);
    }
    formData.append("street_address", venueData.street_address);
    formData.append("city", venueData.city);
    formData.append("state", venueData.state);
    formData.append("zip_code", venueData.zip_code);
    if (venueData.capacity !== null && venueData.capacity !== undefined) {
      formData.append("capacity", venueData.capacity);
    }
    formData.append("has_sound_provided", venueData.has_sound_provided);
    formData.append("has_parking", venueData.has_parking);
    if (venueData.age_restriction !== null && venueData.age_restriction !== undefined) {
      formData.append("age_restriction", venueData.age_restriction);
    }
    
    // Add image file if provided
    if (imageFile) {
      formData.append("image", imageFile);
    }

    // For FormData, don't set Content-Type header (browser will set it with boundary)
    const response = await apiClient('/venues/', {
      method: 'POST',
      body: formData,
    });

    if (!response.ok) {
      const errorData = await response.json();
      // Handle validation errors - FastAPI returns detail as array or string
      let errorMessage = "Failed to create venue";
      if (errorData.detail) {
        if (Array.isArray(errorData.detail)) {
          // Validation errors come as an array
          errorMessage = errorData.detail.map(err => {
            const field = err.loc ? err.loc.join('.') : 'field';
            const msg = err.msg || err;
            return `${field}: ${msg}`;
          }).join(", ");
        } else {
          errorMessage = errorData.detail;
        }
      }
      throw new Error(errorMessage);
    }

    return await response.json();
  },

  async joinVenueWithInvite(inviteCode) {
    const response = await apiClient('/venues/join', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        invite_code: inviteCode,
      }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to join venue');
    }

    return await response.json();
  },

  async getUserVenues() {
    // Add timeout to prevent hanging
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 8000); // 8 second timeout
    
    try {
      const response = await apiClient('/venues/my-venues', {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        const error = await response.json();
        const errorMessage = typeof error.detail === 'string' 
          ? error.detail 
          : (error.detail?.message || JSON.stringify(error.detail) || "Failed to fetch venues");
        const err = new Error(errorMessage);
        err.status = response.status;
        throw err;
      }

      return await response.json();
    } catch (error) {
      clearTimeout(timeoutId);
      if (error.name === 'AbortError') {
        const timeoutErr = new Error(
          "Request timeout - venue service may be slow or unavailable"
        );
        timeoutErr.status = 408;
        timeoutErr.isTimeout = true;
        throw timeoutErr;
      }
      throw error;
    }
  },

  async getVenueDetails(venueId) {
    const response = await apiClient(`/venues/${venueId}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to fetch venue details');
    }

    return await response.json();
  },

  async updateVenue(venueId, venueData) {
    const response = await apiClient(`/venues/${venueId}`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(venueData),
    });

    if (!response.ok) {
      const errorData = await response.json();
      let errorMessage = "Failed to update venue";
      if (errorData.detail) {
        if (Array.isArray(errorData.detail)) {
          errorMessage = errorData.detail.map(err => {
            const field = err.loc ? err.loc.join('.') : 'field';
            const msg = err.msg || err;
            return `${field}: ${msg}`;
          }).join(", ");
        } else {
          errorMessage = errorData.detail;
        }
      }
      throw new Error(errorMessage);
    }

    return await response.json();
  },

  async updateVenueImage(venueId, imageFile) {
    const formData = new FormData();
    formData.append("image", imageFile);

    // For FormData, don't set Content-Type header (browser will set it with boundary)
    const response = await apiClient(`/venues/${venueId}/image`, {
      method: 'POST',
      body: formData,
    });

    if (!response.ok) {
      const errorData = await response.json();
      let errorMessage = "Failed to update venue image";
      if (errorData.detail) {
        if (Array.isArray(errorData.detail)) {
          errorMessage = errorData.detail.map(err => {
            const field = err.loc ? err.loc.join('.') : 'field';
            const msg = err.msg || err;
            return `${field}: ${msg}`;
          }).join(", ");
        } else {
          errorMessage = errorData.detail;
        }
      }
      throw new Error(errorMessage);
    }

    return await response.json();
  },

  async getVenueOperatingHours(venueId) {
    const response = await apiClient(`/venues/${venueId}/operating-hours`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to fetch venue operating hours');
    }

    return await response.json();
  },

  async updateVenueOperatingHours(venueId, hoursData) {
    const response = await apiClient(`/venues/${venueId}/operating-hours`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(hoursData),
    });

    if (!response.ok) {
      const errorData = await response.json();
      let errorMessage = "Failed to update venue operating hours";
      if (errorData.detail) {
        if (Array.isArray(errorData.detail)) {
          errorMessage = errorData.detail.map(err => {
            const field = err.loc ? err.loc.join('.') : 'field';
            const msg = err.msg || err;
            return `${field}: ${msg}`;
          }).join(", ");
        } else {
          errorMessage = errorData.detail;
        }
      }
      throw new Error(errorMessage);
    }

    return await response.json();
  },

  async getVenueStaff(venueId) {
    const response = await apiClient(`/venues/${venueId}/staff`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to fetch venue staff');
    }

    return await response.json();
  },

  async listVenues(params = {}) {
    const queryParams = new URLSearchParams();
    
    // Existing filter parameters
    if (params.city) {
      queryParams.append("city", params.city);
    }
    if (params.state) {
      queryParams.append("state", params.state);
    }
    if (params.has_sound_provided !== undefined) {
      queryParams.append("has_sound_provided", params.has_sound_provided);
    }
    if (params.has_parking !== undefined) {
      queryParams.append("has_parking", params.has_parking);
    }
    if (params.min_capacity !== undefined) {
      queryParams.append("min_capacity", params.min_capacity);
    }
    if (params.max_capacity !== undefined) {
      queryParams.append("max_capacity", params.max_capacity);
    }
    if (params.skip !== undefined) {
      queryParams.append("skip", params.skip);
    }
    if (params.limit !== undefined) {
      queryParams.append("limit", params.limit);
    }
    if (params.band_id !== undefined) {
      queryParams.append("band_id", params.band_id);
    }
    
    // Distance filter parameters
    if (params.distance_km !== undefined && params.distance_km !== null) {
      queryParams.append("distance_km", params.distance_km);
    }
    if (params.base_location !== undefined && params.base_location !== null) {
      queryParams.append("base_location", params.base_location);
    }

    const queryString = queryParams.toString();
    const url = `/venues/${queryString ? `?${queryString}` : ""}`;
    
    const response = await apiClient(url, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to fetch venues');
    }

    return await response.json();
  },

  // Venue Equipment Methods
  async getVenueEquipmentCategories(venueId) {
    const response = await apiClient(`/venues/${venueId}/equipment/categories`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to fetch venue equipment categories');
    }

    return await response.json();
  },

  async getVenueEquipment(venueId) {
    const response = await apiClient(`/venues/${venueId}/equipment`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to fetch venue equipment');
    }

    return await response.json();
  },

  async createVenueEquipment(venueId, equipmentData) {
    const response = await apiClient(`/venues/${venueId}/equipment`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(equipmentData),
    });

    if (!response.ok) {
      const errorData = await response.json();
      let errorMessage = "Failed to create venue equipment";
      if (errorData.detail) {
        if (Array.isArray(errorData.detail)) {
          errorMessage = errorData.detail.map(err => {
            const field = err.loc ? err.loc.join('.') : 'field';
            const msg = err.msg || err;
            return `${field}: ${msg}`;
          }).join(", ");
        } else {
          errorMessage = errorData.detail;
        }
      }
      throw new Error(errorMessage);
    }

    return await response.json();
  },

  async updateVenueEquipment(venueId, equipmentId, equipmentData) {
    const response = await apiClient(`/venues/${venueId}/equipment/${equipmentId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(equipmentData),
    });

    if (!response.ok) {
      const errorData = await response.json();
      let errorMessage = "Failed to update venue equipment";
      if (errorData.detail) {
        if (Array.isArray(errorData.detail)) {
          errorMessage = errorData.detail.map(err => {
            const field = err.loc ? err.loc.join('.') : 'field';
            const msg = err.msg || err;
            return `${field}: ${msg}`;
          }).join(", ");
        } else {
          errorMessage = errorData.detail;
        }
      }
      throw new Error(errorMessage);
    }

    return await response.json();
  },

  async deleteVenueEquipment(venueId, equipmentId) {
    const response = await apiClient(`/venues/${venueId}/equipment/${equipmentId}`, {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to delete venue equipment');
    }

    return null; // 204 No Content
  },
};
